File: libraries/typescript/src/providers/xai-voices.ts
/**
 * xAI (Grok) built-in voice catalog.
 *
 * Shared by the TTS, Realtime, and STT providers so the roster, the default
 * voice, and the wire-normalization rule live in exactly one place instead of
 * being duplicated (and drifting) per adapter.
 *
 * xAI's built-in voice ids are lowercase and case-insensitive. A caller may
 * also pass an arbitrary custom cloned `voice_id` (see `xaiCreateCustomVoice`
 * in `./xai-tts`) — this module must never reject those, so
 * {@link isXaiBuiltinVoice} and {@link normalizeXaiVoice} are permissive by
 * design: anything that isn't a recognized built-in id is left alone.
 *
 * Source: https://docs.x.ai/developers/model-capabilities/audio/voice
 */

/** xAI's default built-in voice. */
export const XAI_DEFAULT_VOICE = 'eve';

/** All 26 built-in xAI voice ids, alphabetical. */
export const XAI_VOICE_IDS = [
  'altair',
  'ara',
  'atlas',
  'carina',
  'castor',
  'celeste',
  'cosmo',
  'eve',
  'helios',
  'helix',
  'iris',
  'kepler',
  'leo',
  'lumen',
  'luna',
  'lux',
  'naksh',
  'orion',
  'perseus',
  'rex',
  'rigel',
  'sal',
  'sirius',
  'ursa',
  'zagan',
  'zenith',
] as const;

/** One of xAI's 26 built-in voice ids (always lowercase). */
export type XaiVoiceId = (typeof XAI_VOICE_IDS)[number];

/** Tone and suggested use cases for one built-in xAI voice. */
export interface XaiVoiceInfo {
  readonly id: XaiVoiceId;
  readonly tone: string;
  readonly useCases: readonly string[];
}

/**
 * Full built-in voice roster in xAI's documented display order — `eve` (the
 * default) first, then the rest alphabetically. Tone and use cases are from
 * the xAI Voice Overview (docs.x.ai/developers/model-capabilities/audio/voice).
 */
export const XAI_VOICES: readonly XaiVoiceInfo[] = [
  { id: 'eve', tone: 'Energetic and upbeat', useCases: [] },
  {
    id: 'altair',
    tone: 'Elegant, refined, and effortlessly premium',
    useCases: ['Advertising', 'Narration'],
  },
  { id: 'ara', tone: 'Warm and friendly', useCases: [] },
  {
    id: 'atlas',
    tone: 'Confident, commanding, and reassuring',
    useCases: ['Sales', 'Assistant'],
  },
  {
    id: 'carina',
    tone: 'Soft, empathetic, and soothing',
    useCases: ['Wellness', 'Support'],
  },
  {
    id: 'castor',
    tone: 'Charismatic, down-to-earth, and easygoing',
    useCases: ['Sales', 'Support'],
  },
  {
    id: 'celeste',
    tone: 'Compassionate, confident, and reassuring',
    useCases: ['Support', 'Assistant'],
  },
  {
    id: 'cosmo',
    tone: 'Bright, curious, and easy to follow',
    useCases: ['Education', 'Podcast'],
  },
  {
    id: 'helios',
    tone: 'Upbeat, energetic, and endlessly versatile',
    useCases: ['Assistant', 'Wellness'],
  },
  {
    id: 'helix',
    tone: 'Bold, dynamic, and adrenaline-fueled',
    useCases: ['Commentary', 'Podcast'],
  },
  {
    id: 'iris',
    tone: 'Friendly, upbeat, and naturally charming',
    useCases: ['Sales', 'Support'],
  },
  {
    id: 'kepler',
    tone: 'Inventive, forward-thinking, and charismatic',
    useCases: ['Advertising', 'Podcast'],
  },
  { id: 'leo', tone: 'Authoritative and strong', useCases: [] },
  {
    id: 'lumen',
    tone: 'Warm, articulate, and engaging',
    useCases: ['Education', 'Advertising'],
  },
  {
    id: 'luna',
    tone: 'Gentle, patient, and deeply nurturing',
    useCases: ['Education', 'Assistant'],
  },
  {
    id: 'lux',
    tone: 'Grounded, calm, and quietly wise',
    useCases: ['Wellness', 'Narration'],
  },
  {
    id: 'naksh',
    tone: 'Warm, thoughtful, and wise',
    useCases: ['Assistant', 'Support'],
  },
  {
    id: 'orion',
    tone: 'Rich, cinematic, and resonant',
    useCases: ['Narration', 'Audiobooks'],
  },
  {
    id: 'perseus',
    tone: 'Strong, confident, and trustworthy',
    useCases: ['Advertising', 'Narration'],
  },
  { id: 'rex', tone: 'Confident and clear', useCases: [] },
  {
    id: 'rigel',
    tone: 'Precise, professional, and calmly confident',
    useCases: ['Assistant', 'Support'],
  },
  { id: 'sal', tone: 'Smooth and balanced', useCases: [] },
  {
    id: 'sirius',
    tone: 'Quick-witted, clever, and playful',
    useCases: ['Commentary', 'Characters'],
  },
  {
    id: 'ursa',
    tone: 'Friendly, warm, and steadfast',
    useCases: ['Assistant', 'Podcast'],
  },
  {
    id: 'zagan',
    tone: 'Powerful, dramatic, and unmistakable',
    useCases: ['Characters', 'Narration'],
  },
  {
    id: 'zenith',
    tone: 'Sharp, focused, and driven',
    useCases: ['Sales', 'Advertising'],
  },
];

/** Lookup set backing {@link isXaiBuiltinVoice} / {@link normalizeXaiVoice}. */
const BUILTIN_VOICE_SET: ReadonlySet<string> = new Set(XAI_VOICE_IDS);

/**
 * True when `voice` (trimmed, case-insensitive) names one of xAI's 26
 * built-in voices. `false` covers everything else, INCLUDING a legitimate
 * custom cloned `voice_id` — callers must treat `false` as "not a documented
 * preset", never as "invalid".
 */
export function isXaiBuiltinVoice(voice: string): voice is XaiVoiceId {
  return BUILTIN_VOICE_SET.has(voice.trim().toLowerCase());
}

/**
 * Normalize a voice id for the wire: trim whitespace, then lowercase ONLY
 * when it names a built-in voice (xAI's built-in ids are case-insensitive).
 * A custom cloned `voice_id` is an arbitrary opaque string, so it is trimmed
 * but never case-folded — xAI matches those ids exactly.
 */
export function normalizeXaiVoice(voice: string): string {
  const trimmed = voice.trim();
  const lowered = trimmed.toLowerCase();
  return BUILTIN_VOICE_SET.has(lowered) ? lowered : trimmed;
}

/** Look up tone/use-case metadata for a built-in voice id (case-insensitive). */
export function getXaiVoice(id: string): XaiVoiceInfo | undefined {
  const normalized = id.trim().toLowerCase();
  return XAI_VOICES.find((v) => v.id === normalized);
}

File: libraries/python/getpatter/providers/xai_voices.py
"""xAI built-in voice catalog for the Patter SDK.

Single source of truth for the 26 built-in xAI voice ids, shared by
:class:`~getpatter.providers.xai_tts.XaiTTS` and
:class:`~getpatter.providers.xai_realtime.XaiRealtimeAdapter` (the Voice Agent
API draws on the same roster). Both modules import :data:`XAI_DEFAULT_VOICE`
and :func:`normalize_xai_voice` from here instead of hardcoding them, so the
roster has one place to update.

Built-in ids are lowercase and case-insensitive on the xAI side (docs.x.ai).
Custom cloned voices (see
:func:`~getpatter.providers.xai_tts.create_custom_voice`) return an opaque
``voice_id`` string that is NOT part of this roster — :func:`normalize_xai_voice`
only folds case for a KNOWN built-in id and otherwise leaves the value
untouched (besides trimming whitespace), so a custom voice id is never
rejected or reshaped.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class XaiVoice:
    """Metadata for one built-in xAI voice (from the xAI voice catalog)."""

    id: str
    tone: str
    use_cases: tuple[str, ...]


# Roster order: ``eve`` (the documented default) first, then the remaining 25
# alphabetically by id — mirrors the table in docs/*/providers/xai-tts.mdx.
XAI_VOICES: tuple[XaiVoice, ...] = (
    XaiVoice("eve", "Energetic and upbeat", ()),
    XaiVoice(
        "altair",
        "Elegant, refined, and effortlessly premium",
        ("Advertising", "Narration"),
    ),
    XaiVoice("ara", "Warm and friendly", ()),
    XaiVoice(
        "atlas", "Confident, commanding, and reassuring", ("Sales", "Assistant")
    ),
    XaiVoice("carina", "Soft, empathetic, and soothing", ("Wellness", "Support")),
    XaiVoice(
        "castor",
        "Charismatic, down-to-earth, and easygoing",
        ("Sales", "Support"),
    ),
    XaiVoice(
        "celeste",
        "Compassionate, confident, and reassuring",
        ("Support", "Assistant"),
    ),
    XaiVoice("cosmo", "Bright, curious, and easy to follow", ("Education", "Podcast")),
    XaiVoice(
        "helios",
        "Upbeat, energetic, and endlessly versatile",
        ("Assistant", "Wellness"),
    ),
    XaiVoice(
        "helix", "Bold, dynamic, and adrenaline-fueled", ("Commentary", "Podcast")
    ),
    XaiVoice(
        "iris", "Friendly, upbeat, and naturally charming", ("Sales", "Support")
    ),
    XaiVoice(
        "kepler",
        "Inventive, forward-thinking, and charismatic",
        ("Advertising", "Podcast"),
    ),
    XaiVoice("leo", "Authoritative and strong", ()),
    XaiVoice(
        "lumen", "Warm, articulate, and engaging", ("Education", "Advertising")
    ),
    XaiVoice(
        "luna", "Gentle, patient, and deeply nurturing", ("Education", "Assistant")
    ),
    XaiVoice("lux", "Grounded, calm, and quietly wise", ("Wellness", "Narration")),
    XaiVoice("naksh", "Warm, thoughtful, and wise", ("Assistant", "Support")),
    XaiVoice("orion", "Rich, cinematic, and resonant", ("Narration", "Audiobooks")),
    XaiVoice(
        "perseus",
        "Strong, confident, and trustworthy",
        ("Advertising", "Narration"),
    ),
    XaiVoice("rex", "Confident and clear", ()),
    XaiVoice(
        "rigel",
        "Precise, professional, and calmly confident",
        ("Assistant", "Support"),
    ),
    XaiVoice("sal", "Smooth and balanced", ()),
    XaiVoice(
        "sirius", "Quick-witted, clever, and playful", ("Commentary", "Characters")
    ),
    XaiVoice("ursa", "Friendly, warm, and steadfast", ("Assistant", "Podcast")),
    XaiVoice(
        "zagan",
        "Powerful, dramatic, and unmistakable",
        ("Characters", "Narration"),
    ),
    XaiVoice("zenith", "Sharp, focused, and driven", ("Sales", "Advertising")),
)

XAI_VOICE_IDS: tuple[str, ...] = tuple(v.id for v in XAI_VOICES)

# xAI's documented default voice. Derived from the roster (``XAI_VOICES[0]``)
# rather than duplicated as a second literal, so the "eve first" ordering
# invariant above is the single source of truth.
XAI_DEFAULT_VOICE: str = XAI_VOICES[0].id

# Case-insensitive lookup built once from the roster above.
_VOICES_BY_ID: dict[str, XaiVoice] = {v.id.lower(): v for v in XAI_VOICES}


def is_xai_builtin_voice(voice: str) -> bool:
    """Return True if *voice* names one of the built-in roster voices.

    Case-insensitive and trims surrounding whitespace, matching how xAI
    resolves the ``voice`` / ``voice_id`` field on its API. A custom cloned
    voice id returns False here — it is still a legal value, just not one
    with roster metadata.
    """
    return voice.strip().lower() in _VOICES_BY_ID


def normalize_xai_voice(voice: str) -> str:
    """Prepare a ``voice`` value for the xAI wire format.

    Built-in roster ids are lowercase and case-insensitive on the xAI side, so
    a known id is trimmed and lowercased before being sent. A custom cloned
    ``voice_id`` (see :func:`~getpatter.providers.xai_tts.create_custom_voice`)
    is an arbitrary, case-sensitive opaque string — it must NEVER be rejected
    or reshaped, so only surrounding whitespace is trimmed.
    """
    trimmed = voice.strip()
    lowered = trimmed.lower()
    return lowered if lowered in _VOICES_BY_ID else trimmed


def get_xai_voice(voice_id: str) -> XaiVoice | None:
    """Look up roster metadata for a built-in voice id, else ``None``.

    Case-insensitive and trims surrounding whitespace; returns ``None`` for a
    custom cloned voice id since it carries no roster metadata.
    """
    return _VOICES_BY_ID.get(voice_id.strip().lower())
